clash-prelude/src/Clash/Explicit/Signal.hs

```diff
@@ -859,4 +861,70 @@ sampleWithResetN
 sampleWithResetN nReset nSamples f =
   take nSamples (sampleWithReset nReset f)
 
+-- | Simulate a component until it matches a condition
+--
+-- It prints a message of the form
+--
+-- "Signal sampled for N cycles until value X"
+--
+-- A common usage is with a test bench using
+-- 'Clash.Explicit.Testbench.outputVerifier'. __NB__: Since this uses
+-- 'Clash.Explicit.Testbench.assert', when using @clashi@, read the
+-- note at "Clash.Explicit.Testbench#assert-clashi".
+--
+-- __NB__: This function is not synthesizable
+--
+-- === __Example with test bench__
+--
+-- @
+-- import Clash.Prelude
+-- import Clash.Explicit.Testbench
+--
+-- import qualified Prelude as P
+--
+-- topEntity
+--   :: 'Signal' 'System' Int
+--   -> 'Signal' 'System' Int
+-- topEntity = id
+--
+-- testBench
+--   :: 'Signal' 'System' Bool
+-- testBench = done
+--  where
+--   testInput = 'Clash.Explicit.Testbench.stimuliGenerator' clk rst $('Clash.Sized.Vector.listToVecTH' [1 :: Int .. 10])
+--   expectedOutput =
+--     'Clash.Explicit.Testbench.outputVerifier'' clk rst $('Clash.Sized.Vector.listToVecTH' $ [1 :: Int .. 9] P.++ [42])
+--   done = expectedOutput $ topEntity testInput
+--   clk = 'Clash.Explicit.Testbench.tbSystemClockGen' (not \<$\> done)
+--   rst = 'systemResetGen'
+-- @
+--
+-- @
+-- > runUntil id testBench
+--
+--
+-- cycle(\<Clock: System\>): 10, outputVerifier
+-- expected value: 42, not equal to actual value: 10
+-- Signal sampled for 11 cycles until value True
+-- @
+
+-- The "value `seqX`" bit is to have any invocations of 'trace', such as in
+-- our 'assert' (, 'outputVerifier', ...) functions, print before printing the
+-- result message.
+runUntil
+  :: forall dom a
+   . (KnownDomain dom, NFDataX a, ShowX a)
+  => (a -> Bool)
+  -- ^ Condition checking function, should return @True@ to finish run
+  -> Signal dom a
+  -- ^ 'Signal' we want to sample for the condition
+  -> IO ()
+runUntil check s = value `seqX` putStrLn msg
```

Comment:
Perhaps `sampleUntil`? Given that we already have `sample` and `sampleN`.

Reply:
Oh that's nice, +1

Reply:
But `sample` and `sampleN` return samples, `runUntl` doesn't return them. I agree that it does sample the signal, to be able to do the testing, but if it were called `sampleUntil` I would think it would return the list of samples until the test was true.

Reply:
Although I guess that'd mean it'd have to return `[a]`, so perhaps this is better after all.

Reply:
It could be worth having both? I can see a `sampleUntil` that returns the values also being useful

Reply:
I always have Haskell's `Prelude` imported qualified under `P` in my `clashi` sessions, and then it's as simple as
```
>>> P.takeWhile not $ sample testBench
```
to get what I imagine would look like
```
>>> sampleUntil id testBench
```
with that new `sampleUntil`. To me, the former is just as good and I wouldn't need the latter.

Reply:
(although I suppose if I were to design `sampleUntil`, I would include the first sample passing the check as well, so it would not be exactly equivalent)

Reply:
I suddenly realise I've completely forgotten to add some nice invocations to the `runUntil` example. To wit,
```
>>> mapM_ runUntil [ testBenchA, testBenchB ]
>>> sequence_ [ runUntil testBenchA, runUntil testBenchB ]
```
The latter for when the clock domain of the two testbenches is not the same.

Reply:
For a single test bench, `P.takeWhile not $ sample testBench` is pretty okay already, it's when you have multiple test benches that `runUntil` becomes a prety nice shorthand. So I had always intended to steer the reader towards that kind of use, and then totally forgot to put it in once I was actually writing the doc...

Reply:
> I always have Haskell's `Prelude` imported qualified under `P` in my `clashi` sessions

To take a harsh stance: if you need to have `Prelude` from `base` imported in `clashi`, `Clash.Prelude` isn't doing it's job well enough. For me the pattern of `P.takeWhile not $ sample testBench` motivates a `sampleUntil` function being included because that's friendlier to beginners who may not be aware there's a second prelude they can import

Reply:
We are doing enough already... I was not aware, but you can just do
```
> takeWhile not $ sample testBench
```
:-)

Reply:
Hmm okay, so the situation isn't as bad as I thought. In that case I guess it's not so bad to not have `sampleUntil`, although maybe it would still be a nice touch given `sampleN` is basically just a synonym for `take` after `sample` and `sampleUntil` is a synonym for `takeWhile (not . p)` for some `p` after `sample`

I leave it to your own preference :slightly_smiling_face:

Reply:
I think a `sampleUntil` with the documentation pointers you suggested on Slack are a good addition.